File: backend/src/platform/jcs.ts
/**
 * RFC 8785 JSON Canonicalization Scheme.
 *
 * Object keys sorted by UTF-16 code units; no whitespace; numbers use the
 * ECMAScript Number-to-string algorithm (which JSON.stringify already
 * implements); strings escaped as JSON.stringify does.
 */
export function canonicalize(value: unknown): string {
  if (value === null || typeof value === 'boolean') return JSON.stringify(value);
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) throw new TypeError('JCS: non-finite number');
    return JSON.stringify(value);
  }
  if (typeof value === 'string') return JSON.stringify(value);
  if (value instanceof Date) return JSON.stringify(value.toISOString());
  if (Array.isArray(value)) return `[${value.map((v) => canonicalize(v === undefined ? null : v)).join(',')}]`;
  if (typeof value === 'object') {
    const entries = Object.entries(value as Record<string, unknown>)
      .filter(([, v]) => v !== undefined)
      // Default sort compares UTF-16 code units, as RFC 8785 requires.
      .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
    return `{${entries.map(([k, v]) => `${JSON.stringify(k)}:${canonicalize(v)}`).join(',')}}`;
  }
  throw new TypeError(`JCS: unsupported type ${typeof value}`);
}

File: app/lib/src/jcs.dart
import 'dart:convert';

/// RFC 8785 JSON Canonicalization Scheme (ADR-0011).
///
/// The app signs these exact bytes and sends them as-is; the server verifies
/// over the bytes it received and never re-serializes, so this only has to be
/// deterministic. It follows JCS anyway: object keys sorted by UTF-16 code
/// units, no whitespace, integral numbers without a fraction.
String canonicalize(Object? v) {
  if (v == null) return 'null';
  if (v is bool) return v ? 'true' : 'false';
  if (v is int) return v.toString();
  if (v is double) {
    if (!v.isFinite) throw ArgumentError('JCS: non-finite number');
    if (v == v.truncateToDouble() && v.abs() < 1e21) return v.toInt().toString();
    return v.toString();
  }
  if (v is String) return jsonEncode(v);
  if (v is List) return '[${v.map(canonicalize).join(',')}]';
  if (v is Map) {
    final keys = v.keys.map((k) => k as String).toList()..sort(); // String.compareTo = UTF-16 code units
    return '{${keys.where((k) => v[k] != null || v.containsKey(k)).map((k) => '${jsonEncode(k)}:${canonicalize(v[k])}').join(',')}}';
  }
  throw ArgumentError('JCS: unsupported type ${v.runtimeType}');
}

List<int> canonicalBytes(Object? v) => utf8.encode(canonicalize(v));

String b64url(List<int> bytes) => base64Url.encode(bytes).replaceAll('=', '');

List<int> b64urlDecode(String s) => base64Url.decode(s.padRight((s.length + 3) ~/ 4 * 4, '='));
